### ISSUE TITLE
@JsonUnwrapped ignored with (embedded) CollectionModel and HAL Forms

### ISSUE THREAD

**Issue by toedter** [spring-projects/spring-hateoas#968]
To create REST resources with a list of embedded resources I use a pattern like

```
@JsonUnwrapped
CollectionModel<EntityModel<Movie>> movies;

```
As an example:

```
@Data
public class DirectorRepresentationModel extends RepresentationModel {
    private long id = 1;
    private String name = "Steven Spielberg";

    @JsonUnwrapped
    CollectionModel<EntityModel<Movie>> movies;

    public DirectorRepresentationModel() {
        Movie movie1 = new Movie();
        movie1.setId("1");
        movie1.setTitle("ET");

        Movie movie2 = new Movie();
        movie2.setId("2");
        movie2.setTitle("Jaws");

        EntityModel<Movie> p1 = new EntityModel<>(movie1);
        EntityModel<Movie> p2 = new EntityModel<>(movie2);

        movies = new CollectionModel<>(Arrays.asList(p1, p2));
    }
}
```

When setting the hypermedia configuration to HAL, it is rendered as I would expect:

```
{
  id: 1,
  name: "Steven Spielberg",
  _embedded: {
    movieList: [
      {
          id: "1",
          title: "ET"
      },
      {
           id: "2",
           title: "Jaws"
      }
     ]
   }
}
```

But when setting the hypermedia configuration to HAL_FORMS, the @JsonUnwrapped does not apply:

```
{
  id: 1,
  name: "Steven Spielberg",
  movies: {
    _embedded: {
       movieList: [
         {
             id: "1",
             title: "ET"
         },
         {
              id: "2",
              title: "Jaws"
         }
        ]
     }
  }
}
```

IMHO HAL_FORMS should render this exactly like HAL does.

**Comment by gregturn** [comment]
HAL-FORMS serialization copies the model’s data into a set of classes aimed at signaling Jackson on how to render its bits. HAL just serializes the Model directly.  

I had not anticipated mixing model types like you did. Should make the perfect test case!

Additionally this information is useful for #864.

**Comment by toedter** [comment]
Using the latest snapshot (March 31.) HAL_FORMS would not render any property in a derived RepresentationModel anymore. When setting the hypermedia configuration to HAL_FORMS, the above example is now rendered as:

```
{
  id: 1,
  name: "Steven Spielberg"
}
```

**Comment by gregturn** [comment]
> IMHO it is very important that the HAL_FORMS implementation renders RepresentationModels that don't use any affordances (which will be mapped on HAL Forms _templates) the very same way that the HAL implementation renders them.

Agreed. 

Thought I had a guess what “fix” broke your solution. Turns out I was wrong. 

The point is, HAL and HAL-FORMS use very different serialization strategies so the solution isn’t straight forward. 

To be honest I’d like to rewrite HAL so that it’s just like HAL-FORMS offering better consistency in design. 

We could almost use a TCK...a batch of tests that exercise a wide range of model combinations with expected outcomes.

**Comment by gregturn** [comment]
Related to #970.

**Comment by toedter** [comment]
A workaround for HAL_FORMS rendering representation model properties is to always use the pattern

`public class MyModel extends RepresentationModel<MyModel>`

Then all the attributes in MyModel are treated as _content_.
But still: `@JsonUnwrapped` would not work for

```
@JsonUnwrapped
CollectionModel<EntityModel<Movie>> movies;
```

**Comment by odrotbohm** [comment]
Fixed with #1269.
